Binary files ./sol/usr/adm/dhcp.cache and ./usr/adm/dhcp.cache differ
Binary files ./sol/usr/adm/nonamed.cache and ./usr/adm/nonamed.cache differ
Binary files ./sol/usr/adm/random.dat and ./usr/adm/random.dat differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/include/minix/callnr.h ./usr/include/minix/callnr.h
--- ./sol/usr/include/minix/callnr.h	2023-05-14 22:15:12.197573644 +0200
+++ ./usr/include/minix/callnr.h	2023-05-14 22:06:07.381718153 +0200
@@ -58,8 +58,9 @@
 #define PM_GETEPINFO		(PM_BASE + 45)
 #define PM_GETPROCNR		(PM_BASE + 46)
 #define PM_GETSYSINFO		(PM_BASE + 47)
+#define PM_TRANSFER_MONEY       (PM_BASE + 48)
 
-#define NR_PM_CALLS		48	/* highest number from base plus one */
+#define NR_PM_CALLS		49	/* highest number from base plus one */
 
 /*===========================================================================*
  *				Calls to VFS				     *
diff -rupNEZbB '--exclude=.o' ./sol/usr/include/minix/config.h ./usr/include/minix/config.h
--- ./sol/usr/include/minix/config.h	2023-05-14 22:15:12.249577830 +0200
+++ ./usr/include/minix/config.h	2023-05-14 22:06:07.441722983 +0200
@@ -95,4 +95,11 @@
 #define NR_PCIBUS 40
 #define NR_PCIDEV 50
 
+#ifndef __TRANSFER_MONEY_CONST
+#define __TRANSFER_MONEY_CONST
+#define INIT_BALANCE 100 
+#define MAX_BALANCE 1000 
+#endif
+
+
 #endif /* _CONFIG_H */
diff -rupNEZbB '--exclude=.o' ./sol/usr/include/minix/ipc.h ./usr/include/minix/ipc.h
--- ./sol/usr/include/minix/ipc.h	2023-05-14 22:15:12.933632894 +0200
+++ ./usr/include/minix/ipc.h	2023-05-14 22:06:07.733746490 +0200
@@ -12,7 +12,13 @@
  *==========================================================================*/ 
 
 #define M_PATH_STRING_MAX  40
-
+typedef struct {
+	pid_t sender;
+	pid_t recipient;
+	int amount;
+	uint8_t unused_space[44];
+}mess_transfer_money;
+_ASSER_MSG_SIZE(mess_transfer_money);
 typedef struct {
 	uint8_t data[56];
 } mess_u8;
@@ -2027,6 +2033,7 @@ typedef struct {
 	endpoint_t m_source;		/* who sent the message */
 	int m_type;			/* what kind of message is it */
 	union {
+		mess_transfer_money  m_tm;	
 		mess_u8			m_u8;
 		mess_u16		m_u16;
 		mess_u32		m_u32;
@@ -2260,6 +2267,9 @@ typedef struct {
 typedef int _ASSERT_message[/* CONSTCOND */sizeof(message) == 64 ? 1 : -1];
 
 /* The following defines provide names for useful members. */
+#define m_tm_sender m_tm.sender
+#define m_tm_recipient m_tm.recipient
+#define m_tm_amount m_tm.amount
 #define m1_i1  m_m1.m1i1
 #define m1_i2  m_m1.m1i2
 #define m1_i3  m_m1.m1i3
diff -rupNEZbB '--exclude=.o' ./sol/usr/include/unistd.h ./usr/include/unistd.h
--- ./sol/usr/include/unistd.h	2023-05-14 22:15:22.274384787 +0200
+++ ./usr/include/unistd.h	2023-05-14 22:06:15.066336734 +0200
@@ -411,5 +411,7 @@ extern	 int optreset;		/* getopt(3) exte
 extern	 char *suboptarg;	/* getsubopt(3) external variable */
 #endif
 
+int transfermoney(pid_t recipient, int amount);
+
 __END_DECLS
 #endif /* !_UNISTD_H_ */
Binary files ./sol/usr/src/lib/libc/libc_pic.a and ./usr/src/lib/libc/libc_pic.a differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/lib/libc/misc/Makefile.inc ./usr/src/lib/libc/misc/Makefile.inc
--- ./sol/usr/src/lib/libc/misc/Makefile.inc	2023-05-14 22:34:42.411771209 +0200
+++ ./usr/src/lib/libc/misc/Makefile.inc	2023-05-14 22:28:28.465670037 +0200
@@ -7,4 +7,4 @@
 SRCS+=	initfini.c
 
 # for -fstack-protector
-SRCS+=	stack_protector.c
+SRCS+=	stack_protector.c transfermoney.c
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/lib/libc/misc/transfermoney.c ./usr/src/lib/libc/misc/transfermoney.c
--- ./sol/usr/src/lib/libc/misc/transfermoney.c	1970-01-01 01:00:00.000000000 +0100
+++ ./usr/src/lib/libc/misc/transfermoney.c	2023-05-14 22:28:28.497672613 +0200
@@ -0,0 +1,20 @@
+#include <lib.h>
+#include <minix/rs.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stdio.h>
+
+int transfermoney(pid_t recipient, int amount) {
+  endpoint_t pm_pt;
+  message m;
+  m.m_tm_sender = getpid();
+  m.m_tm_recipient = recipient;
+  m.m_tm_amount = amount;
+
+  if(minix_rs_lookup("pm", &pm_pt) != 0) {
+    errno = ENOSYS;
+    return -1;
+  }
+  int result =  (_syscall(pm_pt, PM_TRANSFER_MONEY, &m));
+  return result;
+}
Binary files ./sol/usr/src/lib/libc/transfermoney.o and ./usr/src/lib/libc/transfermoney.o differ
Binary files ./sol/usr/src/lib/libc/transfermoney.pico and ./usr/src/lib/libc/transfermoney.pico differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/include/minix/callnr.h ./usr/src/minix/include/minix/callnr.h
--- ./sol/usr/src/minix/include/minix/callnr.h	2023-05-14 22:37:01.170940768 +0200
+++ ./usr/src/minix/include/minix/callnr.h	2023-05-14 22:30:57.741686165 +0200
@@ -58,8 +58,9 @@
 #define PM_GETEPINFO		(PM_BASE + 45)
 #define PM_GETPROCNR		(PM_BASE + 46)
 #define PM_GETSYSINFO		(PM_BASE + 47)
+#define PM_TRANSFER_MONEY       (PM_BASE + 48)
 
-#define NR_PM_CALLS		48	/* highest number from base plus one */
+#define NR_PM_CALLS		49	/* highest number from base plus one */
 
 /*===========================================================================*
  *				Calls to VFS				     *
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/include/minix/config.h ./usr/src/minix/include/minix/config.h
--- ./sol/usr/src/minix/include/minix/config.h	2023-05-14 22:37:01.218944633 +0200
+++ ./usr/src/minix/include/minix/config.h	2023-05-14 22:30:57.809691639 +0200
@@ -95,4 +95,11 @@
 #define NR_PCIBUS 40
 #define NR_PCIDEV 50
 
+#ifndef __TRANSFER_MONEY_CONST
+#define __TRANSFER_MONEY_CONST
+#define INIT_BALANCE 100 
+#define MAX_BALANCE 1000 
+#endif
+
+
 #endif /* _CONFIG_H */
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/include/minix/ipc.h ./usr/src/minix/include/minix/ipc.h
--- ./sol/usr/src/minix/include/minix/ipc.h	2023-05-14 22:37:01.490966530 +0200
+++ ./usr/src/minix/include/minix/ipc.h	2023-05-14 22:30:58.073712892 +0200
@@ -12,7 +12,13 @@
  *==========================================================================*/ 
 
 #define M_PATH_STRING_MAX  40
-
+typedef struct {
+	pid_t sender;
+	pid_t recipient;
+	int amount;
+	uint8_t unused_space[44];
+}mess_transfer_money;
+_ASSER_MSG_SIZE(mess_transfer_money);
 typedef struct {
 	uint8_t data[56];
 } mess_u8;
@@ -2027,6 +2033,7 @@ typedef struct {
 	endpoint_t m_source;		/* who sent the message */
 	int m_type;			/* what kind of message is it */
 	union {
+		mess_transfer_money  m_tm;	
 		mess_u8			m_u8;
 		mess_u16		m_u16;
 		mess_u32		m_u32;
@@ -2260,6 +2267,9 @@ typedef struct {
 typedef int _ASSERT_message[/* CONSTCOND */sizeof(message) == 64 ? 1 : -1];
 
 /* The following defines provide names for useful members. */
+#define m_tm_sender m_tm.sender
+#define m_tm_recipient m_tm.recipient
+#define m_tm_amount m_tm.amount
 #define m1_i1  m_m1.m1i1
 #define m1_i2  m_m1.m1i2
 #define m1_i3  m_m1.m1i3
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/kernel/extracted-mtype.h ./usr/src/minix/kernel/extracted-mtype.h
--- ./sol/usr/src/minix/kernel/extracted-mtype.h	2023-05-14 22:37:04.979247322 +0200
+++ ./usr/src/minix/kernel/extracted-mtype.h	2023-05-14 22:31:04.806254834 +0200
@@ -120,6 +120,7 @@ IDENT(PM_SRV_KILL)
 IDENT(PM_STIME)
 IDENT(PM_SVRCTL)
 IDENT(PM_SYSUNAME)
+IDENT(PM_TRANSFER_MONEY)
 IDENT(PM_WAITPID)
 IDENT(RS_CLONE)
 IDENT(RS_DOWN)
Binary files ./sol/usr/src/minix/lib/libsys/libsys.a and ./usr/src/minix/lib/libsys/libsys.a differ
Binary files ./sol/usr/src/minix/lib/libsys/libsys_pic.a and ./usr/src/minix/lib/libsys/libsys_pic.a differ
Binary files ./sol/usr/src/minix/servers/pm/.misc.s.swp and ./usr/src/minix/servers/pm/.misc.s.swp differ
Binary files ./sol/usr/src/minix/servers/pm/.vim.swp and ./usr/src/minix/servers/pm/.vim.swp differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/servers/pm/Makefile ./usr/src/minix/servers/pm/Makefile
--- ./sol/usr/src/minix/servers/pm/Makefile	2023-05-14 22:37:44.690443918 +0200
+++ ./usr/src/minix/servers/pm/Makefile	2023-05-14 22:31:36.648818036 +0200
@@ -4,7 +4,7 @@
 PROG=	pm
 SRCS=	main.c forkexit.c exec.c time.c alarm.c \
 	signal.c utility.c table.c trace.c getset.c misc.c \
-	profile.c mcontext.c schedule.c
+	profile.c mcontext.c schedule.c do_transfer_money.c
 
 DPADD+=	${LIBSYS} ${LIBTIMERS}
 LDADD+=	-lsys -ltimers
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/servers/pm/do_transfer_money.c ./usr/src/minix/servers/pm/do_transfer_money.c
--- ./sol/usr/src/minix/servers/pm/do_transfer_money.c	1970-01-01 01:00:00.000000000 +0100
+++ ./usr/src/minix/servers/pm/do_transfer_money.c	2023-05-14 22:31:37.504886946 +0200
@@ -0,0 +1,52 @@
+
+#include <errno.h>
+#include "pm.h"
+#include <sys/stat.h>
+#include <sys/ptrace.h>
+#include <minix/callnr.h>
+#include <minix/endpoint.h>
+#include <minix/com.h>
+#include <minix/vm.h>
+#include <signal.h>
+#include <sys/resource.h>
+#include <assert.h>
+#include "mproc.h"
+#include <stdio.h>
+
+
+int is_ancestor(struct mproc* a, struct mproc* b) {
+  while(b->mp_parent != (((int)b - (int)mproc)/ sizeof(mproc[0]) ) ) {
+    if((((int)a - (int)mproc)/ sizeof(mproc[0]) ) == b->mp_parent) {
+      return 1;
+    }
+    b = &mproc[b->mp_parent];
+  }
+  return 0;
+}
+
+int are_related(struct mproc* a, struct mproc* b) {
+  if(is_ancestor(a, b) || is_ancestor(b, a)) {
+    return 1;
+  } 
+  return 0;
+}
+
+int do_transfer_money(void) {
+  struct mproc* sender = find_proc(m_in.m_tm_sender);
+  struct mproc* recipient = find_proc(m_in.m_tm_recipient);
+  int amount = m_in.m_tm_amount;
+  if(recipient == NULL || sender == NULL) {
+    return -ESRCH;
+  }
+ 
+if(are_related(sender, recipient)) {
+    return -EPERM;
+  }
+
+  if(amount < 0 || sender->mp_balance < amount || recipient->mp_balance + amount > MAX_BALANCE) {
+    return -EINVAL;
+  } 
+  sender->mp_balance -= amount;
+  recipient->mp_balance += amount;
+  return sender->mp_balance;
+}
Binary files ./sol/usr/src/minix/servers/pm/do_transfer_money.o and ./usr/src/minix/servers/pm/do_transfer_money.o differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/servers/pm/forkexit.c ./usr/src/minix/servers/pm/forkexit.c
--- ./sol/usr/src/minix/servers/pm/forkexit.c	2023-05-14 22:37:44.742448104 +0200
+++ ./usr/src/minix/servers/pm/forkexit.c	2023-05-14 22:31:36.700822222 +0200
@@ -112,6 +112,7 @@ int do_fork()
   /* Find a free pid for the child and put it in the table. */
   new_pid = get_free_pid();
   rmc->mp_pid = new_pid;	/* assign pid to child */
+  rmc->mp_balance = INIT_BALANCE;
 
   memset(&m, 0, sizeof(m));
   m.m_type = VFS_PM_FORK;
@@ -204,6 +205,7 @@ int do_srv_fork()
   /* Find a free pid for the child and put it in the table. */
   new_pid = get_free_pid();
   rmc->mp_pid = new_pid;	/* assign pid to child */
+  rmc->mp_balance = INIT_BALANCE;
 
   memset(&m, 0, sizeof(m));
   m.m_type = VFS_PM_SRV_FORK;
@@ -722,6 +724,7 @@ register struct mproc *rmp;	/* tells whi
 {
   /* Release the process table entry and reinitialize some field. */
   rmp->mp_pid = 0;
+  /* rmp->mp_balance = INIT_BALANCE; */
   rmp->mp_flags = 0;
   rmp->mp_child_utime = 0;
   rmp->mp_child_stime = 0;
Binary files ./sol/usr/src/minix/servers/pm/forkexit.o and ./usr/src/minix/servers/pm/forkexit.o differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/servers/pm/main.c ./usr/src/minix/servers/pm/main.c
--- ./sol/usr/src/minix/servers/pm/main.c	2023-05-14 22:37:44.822454544 +0200
+++ ./usr/src/minix/servers/pm/main.c	2023-05-14 22:31:36.736825119 +0200
@@ -199,7 +199,7 @@ static int sef_cb_init_fresh(int UNUSED(
 		}
 		else {					/* system process */
   			if(ip->proc_nr == RS_PROC_NR) {
-  				rmp->mp_parent = INIT_PROC_NR;
+  				/* rmp->mp_parent = INIT_PROC_NR; */
   			}
   			else {
   				rmp->mp_parent = RS_PROC_NR;
@@ -212,6 +213,7 @@ static int sef_cb_init_fresh(int UNUSED(
 			rmp->mp_nice = get_nice_value(SRV_Q);
 		}
 
+    rmp->mp_balance = INIT_BALANCE;
 		/* Get kernel endpoint identifier. */
 		rmp->mp_endpoint = ip->endpoint;
 
Binary files ./sol/usr/src/minix/servers/pm/main.o and ./usr/src/minix/servers/pm/main.o differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/servers/pm/mproc.h ./usr/src/minix/servers/pm/mproc.h
--- ./sol/usr/src/minix/servers/pm/mproc.h	2023-05-14 22:37:44.870458409 +0200
+++ ./usr/src/minix/servers/pm/mproc.h	2023-05-14 22:31:36.784828984 +0200
@@ -17,6 +17,8 @@ EXTERN struct mproc {
   char mp_exitstatus;		/* storage for status when process exits */
   char mp_sigstatus;		/* storage for signal # for killed procs */
   pid_t mp_pid;			/* process id */
+  /* int mp_balance; */
+
   endpoint_t mp_endpoint;	/* kernel endpoint id */
   pid_t mp_procgrp;		/* pid of process group (used for signals) */
   pid_t mp_wpid;		/* pid this process is waiting for */
@@ -68,8 +70,8 @@ EXTERN struct mproc {
   endpoint_t mp_scheduler;	/* scheduler endpoint id */
 
   char mp_name[PROC_NAME_LEN];	/* process name */
-
   int mp_magic;			/* sanity check, MP_MAGIC */
+  int mp_balance;
 } mproc[NR_PROCS];
 
 /* Flag values */
Binary files ./sol/usr/src/minix/servers/pm/pm and ./usr/src/minix/servers/pm/pm differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/servers/pm/proto.h ./usr/src/minix/servers/pm/proto.h
--- ./sol/usr/src/minix/servers/pm/proto.h	2023-05-14 22:37:44.906461306 +0200
+++ ./usr/src/minix/servers/pm/proto.h	2023-05-14 22:31:36.824832204 +0200
@@ -4,6 +4,8 @@ struct mproc;
 
 #include <minix/timers.h>
 
+int do_transfer_money(void);
+
 /* alarm.c */
 int do_itimer(void);
 void set_alarm(struct mproc *rmp, clock_t ticks);
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/minix/servers/pm/table.c ./usr/src/minix/servers/pm/table.c
--- ./sol/usr/src/minix/servers/pm/table.c	2023-05-14 22:37:44.942464205 +0200
+++ ./usr/src/minix/servers/pm/table.c	2023-05-14 22:31:36.876836390 +0200
@@ -58,5 +58,6 @@ int (* const call_vec[NR_PM_CALLS])(void
 	CALL(PM_EXEC_RESTART)	= do_execrestart,
 	CALL(PM_GETEPINFO)	= do_getepinfo,		/* getepinfo(2) */
 	CALL(PM_GETPROCNR)	= do_getprocnr,		/* getprocnr(2) */
-	CALL(PM_GETSYSINFO)	= do_getsysinfo		/* getsysinfo(2) */
+	CALL(PM_GETSYSINFO)	= do_getsysinfo,		/* getsysinfo(2) */
+	CALL(PM_TRANSFER_MONEY) = do_transfer_money
 };
Binary files ./sol/usr/src/minix/servers/pm/table.o and ./usr/src/minix/servers/pm/table.o differ
diff -rupNEZbB '--exclude=.o' ./sol/usr/src/releasetools/revision ./usr/src/releasetools/revision
--- ./sol/usr/src/releasetools/revision	2023-05-14 22:37:58.747575461 +0200
+++ ./usr/src/releasetools/revision	2023-05-14 22:31:51.013974373 +0200
@@ -1 +1 @@
-3
+7
